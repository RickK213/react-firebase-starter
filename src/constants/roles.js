export const ROLES = {
  ADMIN: 'ADMIN',
  BASE_USER: 'BASE_USER'
};
