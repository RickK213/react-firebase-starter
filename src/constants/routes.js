export const ROUTES = {
  ADMIN: { label: 'Admin', path: '/admin' },
  ADMIN_DETAILS: { label: null, path: '/admin/:id' },
  ACCOUNT: { label: 'Account', path: '/account' },
  HOME: { label: 'Home', path: '/' },
  PASSWORD_FORGET: {
    label: 'Forgot Password?',
    path: '/account/password-forget'
  },
  SIGN_IN: { label: 'Sign In', path: '/account/sign-in' },
  SIGN_UP: { label: 'Sign Up', path: '/account/sign-up' }
};
