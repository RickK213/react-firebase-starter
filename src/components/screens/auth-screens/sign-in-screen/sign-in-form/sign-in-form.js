import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { ROUTES } from '../../../../../constants/routes';
import { withFirebase } from '../../../../firebase/index';

const INITIAL_STATE = {
  email: '',
  password: '',
  error: null
};

export class SignInFormComponent extends Component {
  static propTypes = {
    firebase: PropTypes.object,
    history: PropTypes.object
  };

  static defaultProps = {
    firebase: null,
    history: null
  };

  constructor(props) {
    super(props);
    this.state = { ...INITIAL_STATE };

    this.handleOnChange = this.handleOnChange.bind(this);
    this.handleOnSubmit = this.handleOnSubmit.bind(this);
  }

  handleOnChange(event) {
    if (!event) return;
    this.setState({ [event.target.name]: event.target.value });
  }

  handleOnSubmit(event) {
    const { email, password } = this.state;
    const { firebase, history } = this.props;

    firebase
      .doSignInWithEmailAndPassword(email, password)
      .then(authenticatedUser => {
        // eslint-disable-next-line no-console
        console.log('authenticatedUser', authenticatedUser);
        this.setState({ ...INITIAL_STATE });
        history.push(ROUTES.HOME.path);
      })
      .catch(error => {
        this.setState({ error });
      });

    event.preventDefault();
  }

  render() {
    const { email, error, password } = this.state;

    const isInvalid = email === '' || password === '';

    return (
      <form onSubmit={this.handleOnSubmit}>
        <input
          name="email"
          type="email"
          value={email}
          onChange={this.handleOnChange}
          placeholder="Email Address"
        />
        <input
          name="password"
          type="password"
          value={password}
          onChange={this.handleOnChange}
          placeholder="Password"
        />
        <button disabled={isInvalid} type="submit">
          Sign In
        </button>

        {error && <p>{error.message}</p>}
      </form>
    );
  }
}

export const SignInForm = withRouter(withFirebase(SignInFormComponent));
