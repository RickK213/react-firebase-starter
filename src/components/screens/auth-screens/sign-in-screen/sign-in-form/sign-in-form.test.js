/* global React, create, shallow */
import { INITIAL_STATE, SignInFormComponent } from './sign-in-form';
import { ROUTES } from '../../../../../constants/routes';

describe('<SignInFormComponent />', () => {
  it('should render correctly', () => {
    const component = create(<SignInFormComponent />).toJSON();
    expect(component).toMatchSnapshot();
  });

  describe('handleOnChange', () => {
    it('should return undefined if event is not set', () => {
      const component = shallow(<SignInFormComponent />);
      const results = component.instance().handleOnChange();
      expect(results).toEqual(undefined);
    });

    it('should call this.setState with the correct arguments if event is set', () => {
      const event = { target: { name: 'testName', value: 'testValue' } };
      const setState = jest.fn();
      const component = shallow(<SignInFormComponent />);
      component.instance().setState = setState;
      component.instance().handleOnChange(event);
      expect(setState).toHaveBeenCalledWith({
        [event.target.name]: event.target.value
      });
    });
  });

  describe('handleOnSubmit', () => {
    it('should call this.setState, history.push, and event.preventDefault() if firebase call succeeds', async () => {
      const push = jest.fn();
      const history = { push };
      const doSignInWithEmailAndPassword = jest
        .fn()
        .mockResolvedValue('default');
      const firebase = { doSignInWithEmailAndPassword };
      const setState = jest.fn();
      const preventDefault = jest.fn();
      const event = { preventDefault };
      const component = shallow(
        <SignInFormComponent firebase={firebase} history={history} />
      );
      component.instance().setState = setState;
      await component.instance().handleOnSubmit(event);
      expect(setState).toHaveBeenCalledWith({ ...INITIAL_STATE });
      expect(push).toHaveBeenCalledWith(ROUTES.HOME.path);
      expect(preventDefault).toHaveBeenCalled();
    });

    it('should call this.setState with the correct arguments and event.preventDefault() if firebase call fails', async () => {
      const testError = 'testError';
      const doSignInWithEmailAndPassword = jest
        .fn()
        .mockRejectedValue(testError);
      const firebase = { doSignInWithEmailAndPassword };
      const setState = jest.fn();
      const preventDefault = jest.fn();
      const event = { preventDefault };
      const component = shallow(<SignInFormComponent firebase={firebase} />);
      component.instance().setState = setState;
      try {
        await component.instance().handleOnSubmit(event);
      } catch (error) {
        expect(setState).toHaveBeenCalledWith({
          ...INITIAL_STATE,
          error: testError
        });
      }
      expect(preventDefault).toHaveBeenCalled();
    });
  });
});
