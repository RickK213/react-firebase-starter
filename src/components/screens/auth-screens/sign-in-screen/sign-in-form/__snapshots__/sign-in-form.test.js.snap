// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignInFormComponent /> should render correctly 1`] = `
<form
  autoComplete="off"
  onSubmit={[Function]}
  style={
    Object {
      "width": "360px",
    }
  }
>
  <input
    autoComplete="off"
    name="email"
    onChange={[Function]}
    placeholder="Email Address"
    style={
      Object {
        "display": "block",
        "marginBottom": "1rem",
        "padding": ".5rem",
        "width": "100%",
      }
    }
    type="email"
    value=""
  />
  <input
    autoComplete="off"
    name="password"
    onChange={[Function]}
    placeholder="Password"
    style={
      Object {
        "display": "block",
        "marginBottom": "1rem",
        "padding": ".5rem",
        "width": "100%",
      }
    }
    type="password"
    value=""
  />
  <button
    disabled={true}
    style={
      Object {
        "padding": ".5rem",
        "textTransform": "uppercase",
      }
    }
    type="submit"
  >
    Sign In
  </button>
</form>
`;
